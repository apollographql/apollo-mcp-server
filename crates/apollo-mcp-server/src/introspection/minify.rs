use apollo_compiler::schema::{ExtendedType, Type};
use regex::Regex;
use std::{collections::HashMap, sync::OnceLock};

pub trait MinifyExt {
    /// Serialize in minified form
    fn minify(&self) -> String;
}

impl MinifyExt for ExtendedType {
    fn minify(&self) -> String {
        match self {
            ExtendedType::Scalar(scalar_type) => minify_scalar(scalar_type),
            ExtendedType::Object(object_type) => minify_object(object_type),
            ExtendedType::Interface(interface_type) => minify_interface(interface_type),
            ExtendedType::Union(union_type) => minify_union(union_type),
            ExtendedType::Enum(enum_type) => minify_enum(enum_type),
            ExtendedType::InputObject(input_object_type) => minify_input_object(input_object_type),
        }
    }
}

fn minify_scalar(scalar_type: &apollo_compiler::schema::ScalarType) -> String {
    shorten_scalar_names(scalar_type.name.as_str()).to_string()
}

fn minify_object(object_type: &apollo_compiler::schema::ObjectType) -> String {
    let fields = minify_fields(&object_type.fields);
    let type_name = format_type_name_with_description(&object_type.name, &object_type.description);
    let interfaces = format_interfaces(&object_type.implements_interfaces);

    if interfaces.is_empty() {
        format!("T:{type_name}:{fields}")
    } else {
        format!("T:{type_name}<{interfaces}>:{fields}")
    }
}

fn minify_interface(interface_type: &apollo_compiler::schema::InterfaceType) -> String {
    let fields = minify_fields(&interface_type.fields);
    let type_name =
        format_type_name_with_description(&interface_type.name, &interface_type.description);
    format!("F:{type_name}:{fields}")
}

fn minify_union(union_type: &apollo_compiler::schema::UnionType) -> String {
    let member_types = union_type
        .members
        .iter()
        .map(|member| member.as_str())
        .collect::<Vec<&str>>()
        .join(",");
    let type_name = format_type_name_with_description(&union_type.name, &union_type.description);
    format!("U:{type_name}:{member_types}")
}

fn minify_enum(enum_type: &apollo_compiler::schema::EnumType) -> String {
    let values = enum_type
        .values
        .keys()
        .map(|value| value.as_str())
        .collect::<Vec<&str>>()
        .join(",");
    let type_name = format_type_name_with_description(&enum_type.name, &enum_type.description);
    format!("E:{type_name}:{values}")
}

fn minify_input_object(input_object_type: &apollo_compiler::schema::InputObjectType) -> String {
    let fields = minify_input_fields(&input_object_type.fields);
    let type_name =
        format_type_name_with_description(&input_object_type.name, &input_object_type.description);
    format!("I:{type_name}:{fields}")
}

// We should only minify directives that assist the LLM in understanding the schema. This included @deprecated
fn minify_directives(directives: &apollo_compiler::ast::DirectiveList) -> String {
    let mut result = String::new();

    static DIRECTIVES_TO_MINIFY: OnceLock<HashMap<&str, &str>> = OnceLock::new();
    let directives_to_minify =
        DIRECTIVES_TO_MINIFY.get_or_init(|| HashMap::from([("deprecated", "D")]));

    for directive in directives.iter() {
        if let Some(minified_name) = directives_to_minify.get(directive.name.as_str()) {
            if !directive.arguments.is_empty() {
                // Since we're only handling @deprecated right now we can just add the reason and minify it
                let reason = directive
                    .arguments
                    .iter()
                    .find(|a| a.name == "reason")
                    .and_then(|a| a.value.as_str())
                    .unwrap_or("No longer supported")
                    .to_string();
                result.push_str(&format!("@{}(\"{}\")", minified_name, reason));
            } else {
                result.push_str(&format!("@{}", minified_name));
            }
        }
    }
    result
}

fn minify_fields(
    fields: &apollo_compiler::collections::IndexMap<
        apollo_compiler::Name,
        apollo_compiler::schema::Component<apollo_compiler::ast::FieldDefinition>,
    >,
) -> String {
    let mut result = String::new();

    for (field_name, field) in fields.iter() {
        // Add description if present
        if let Some(desc) = field.description.as_ref() {
            result.push_str(&format!("\"{}\"", normalize_description(desc)));
        }

        // Add field name
        result.push_str(field_name.as_str());

        // Add arguments if present
        if !field.arguments.is_empty() {
            result.push('(');
            result.push_str(&minify_arguments(&field.arguments));
            result.push(')');
        }

        // Add field type
        result.push(':');
        result.push_str(&type_name(&field.ty));
        result.push_str(&minify_directives(&field.directives));

        result.push(',');
    }

    // Remove trailing comma
    if !result.is_empty() {
        result.pop();
    }

    result
}

fn minify_input_fields(
    fields: &apollo_compiler::collections::IndexMap<
        apollo_compiler::Name,
        apollo_compiler::schema::Component<apollo_compiler::ast::InputValueDefinition>,
    >,
) -> String {
    let mut result = String::new();

    for (field_name, field) in fields.iter() {
        // Add description if present
        if let Some(desc) = field.description.as_ref() {
            result.push_str(&format!("\"{}\"", normalize_description(desc)));
        }

        // Add field name and type
        result.push_str(field_name.as_str());
        result.push(':');
        result.push_str(&type_name(&field.ty));
        result.push_str(&minify_directives(&field.directives));
        result.push(',');
    }

    // Remove trailing comma
    if !result.is_empty() {
        result.pop();
    }

    result
}

fn minify_arguments(
    arguments: &[apollo_compiler::Node<apollo_compiler::ast::InputValueDefinition>],
) -> String {
    arguments
        .iter()
        .map(|arg| {
            if let Some(desc) = arg.description.as_ref() {
                format!(
                    "\"{}\"{}:{}{}",
                    normalize_description(desc),
                    arg.name.as_str(),
                    type_name(&arg.ty),
                    minify_directives(&arg.directives)
                )
            } else {
                format!(
                    "{}:{}{}",
                    arg.name.as_str(),
                    type_name(&arg.ty),
                    minify_directives(&arg.directives)
                )
            }
        })
        .collect::<Vec<String>>()
        .join(",")
}

fn format_type_name_with_description(
    name: &apollo_compiler::Name,
    description: &Option<apollo_compiler::Node<str>>,
) -> String {
    if let Some(desc) = description.as_ref() {
        format!("\"{}\"{}", normalize_description(desc), name)
    } else {
        name.to_string()
    }
}

fn format_interfaces(
    interfaces: &apollo_compiler::collections::IndexSet<apollo_compiler::schema::ComponentName>,
) -> String {
    interfaces
        .iter()
        .map(|interface| interface.as_str())
        .collect::<Vec<&str>>()
        .join(",")
}

fn type_name(ty: &Type) -> String {
    let name = shorten_scalar_names(ty.inner_named_type().as_str());
    if ty.is_list() {
        format!("[{name}]")
    } else if ty.is_non_null() {
        format!("{name}!")
    } else {
        name.to_string()
    }
}

fn shorten_scalar_names(name: &str) -> &str {
    match name {
        "String" => "s",
        "Int" => "i",
        "Float" => "f",
        "Boolean" => "b",
        "ID" => "d",
        _ => name,
    }
}

/// Normalize description formatting
#[allow(clippy::expect_used)]
fn normalize_description(desc: &str) -> String {
    // LLMs can typically process descriptions just fine without whitespace
    static WHITESPACE_PATTERN: OnceLock<Regex> = OnceLock::new();
    let re = WHITESPACE_PATTERN.get_or_init(|| Regex::new(r"\s+").expect("regex pattern compiles"));
    re.replace_all(desc, "").to_string()
}
