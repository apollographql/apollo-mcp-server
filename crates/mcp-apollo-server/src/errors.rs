use apollo_compiler::{Schema, ast::Document, validation::WithErrors};
use rmcp::serde_json;

/// An error in operation parsing
#[derive(Debug, thiserror::Error)]
pub enum OperationError {
    #[error("Could not parse GraphQL document: {0}")]
    GraphQLDocument(WithErrors<Document>),

    #[error("Could not parse GraphQL schema: {0}")]
    GraphQLSchema(WithErrors<Schema>),

    #[error("Internal error: {0}")]
    Internal(String),

    #[error("Operation is missing its required name: {0}")]
    MissingName(String),

    #[error("No operations defined")]
    NoOperations,

    #[error("Invalid JSON: {0}")]
    Json(#[from] serde_json::Error),

    #[error("Too many operations. Expected 1 but got {0}")]
    TooManyOperations(usize),
}

/// An error in server initialization
#[derive(Debug, thiserror::Error)]
pub enum ServerError {
    #[error("Could not parse GraphQL document: {0}")]
    GraphQLDocument(Box<WithErrors<Document>>),

    #[error("Could not parse GraphQL schema: {0}")]
    GraphQLSchema(Box<WithErrors<Schema>>),

    #[error("Invalid JSON: {0}")]
    Json(#[from] serde_json::Error),

    #[error("Failed to create operation: {0}")]
    Operation(#[from] OperationError),

    #[error("Could not open file: {0}")]
    ReadFile(#[from] std::io::Error),
}
