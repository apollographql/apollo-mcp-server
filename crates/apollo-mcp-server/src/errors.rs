use crate::introspection::tools::search::IndexingError;
use apollo_compiler::{Schema, ast::Document, validation::WithErrors};
use apollo_federation::error::FederationError;
use apollo_mcp_registry::platform_api::operation_collections::error::CollectionError;
use reqwest::header::{InvalidHeaderName, InvalidHeaderValue};
use rmcp::serde_json;
use tokio::task::JoinError;
use url::ParseError;

/// An error in operation parsing
#[derive(Debug, thiserror::Error)]
pub enum OperationError {
    #[error("Could not parse GraphQL document: {0}")]
    GraphQLDocument(Box<WithErrors<Document>>),

    #[error("Internal error: {0}")]
    Internal(String),

    #[error("{0}Operation is missing its required name: {1}", .source_path.as_ref().map(|s| format!("{s}: ")).unwrap_or_default(), operation)]
    MissingName {
        source_path: Option<String>,
        operation: String,
    },

    #[error("{0}No operations defined", .source_path.as_ref().map(|s| format!("{s}: ")).unwrap_or_default())]
    NoOperations { source_path: Option<String> },

    #[error("Invalid JSON: {0}")]
    Json(#[from] serde_json::Error),

    #[error("{0}Too many operations. Expected 1 but got {1}", .source_path.as_ref().map(|s| format!("{s}: ")).unwrap_or_default(), count)]
    TooManyOperations {
        source_path: Option<String>,
        count: usize,
    },

    #[error(transparent)]
    File(#[from] std::io::Error),

    #[error("Error loading collection: {0}")]
    Collection(CollectionError),
}

/// An error in server initialization
#[derive(Debug, thiserror::Error)]
pub enum ServerError {
    #[error("Could not parse GraphQL document: {0}")]
    GraphQLDocument(Box<WithErrors<Document>>),

    #[error("Could not parse GraphQL schema: {0}")]
    GraphQLSchema(Box<WithErrors<Schema>>),

    #[error("Could not parse GraphQL schema: {0}")]
    GraphQLDocumentSchema(Box<WithErrors<Document>>),

    #[error("Federation error in GraphQL schema: {0}")]
    Federation(Box<FederationError>),

    #[error("Invalid JSON: {0}")]
    Json(#[from] serde_json::Error),

    #[error("Failed to create operation: {0}")]
    Operation(#[from] OperationError),

    #[error("Could not open file: {0}")]
    ReadFile(#[from] std::io::Error),

    #[error("invalid header value: {0}")]
    HeaderValue(#[from] InvalidHeaderValue),

    #[error("invalid header name: {0}")]
    HeaderName(#[from] InvalidHeaderName),

    #[error("invalid header: {0}")]
    Header(String),

    #[error("invalid custom_scalar_config: {0}")]
    CustomScalarConfig(serde_json::Error),

    #[error("invalid json schema: {0}")]
    CustomScalarJsonSchema(String),

    #[error("Missing environment variable: {0}")]
    EnvironmentVariable(String),

    #[error("You must define operations or enable introspection")]
    NoOperations,

    #[error("No valid schema was supplied")]
    NoSchema,

    #[error("Failed to start server")]
    StartupError(#[from] JoinError),

    #[error("Failed to initialize MCP server")]
    McpInitializeError(#[from] Box<rmcp::service::ServerInitializeError>),

    #[error(transparent)]
    UrlParseError(ParseError),

    #[error("Failed to index schema: {0}")]
    Indexing(#[from] IndexingError),

    #[error("CORS configuration error: {0}")]
    Cors(String),
}

/// An MCP tool error
pub type McpError = rmcp::model::ErrorData;
