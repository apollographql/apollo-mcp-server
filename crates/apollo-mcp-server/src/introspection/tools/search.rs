//! MCP tool to search a GraphQL schema.

use crate::errors::McpError;
use crate::introspection::minify::MinifyExt as _;
use crate::schema_from_type;
use crate::schema_tree_shake::{DepthLimit, SchemaTreeShaker};
use apollo_compiler::ast::{Field, OperationType as AstOperationType, Selection};
use apollo_compiler::validation::Valid;
use apollo_compiler::{Name, Node, Schema};
use apollo_schema_index::{OperationType, Options, SchemaIndex};
use rmcp::model::{CallToolResult, Content, ErrorCode, Tool};
use rmcp::schemars::JsonSchema;
use rmcp::serde_json::Value;
use rmcp::{schemars, serde_json};
use serde::Deserialize;
use std::fmt::Debug;
use std::sync::Arc;
use tokio::sync::Mutex;
use tracing::debug;

/// The name of the tool to search a GraphQL schema.
pub const SEARCH_TOOL_NAME: &str = "search";

/// The maximum number of search results to consider.
const MAX_SEARCH_RESULTS: usize = 5;

/// A tool to search a GraphQL schema.
#[derive(Clone)]
pub struct Search {
    schema: Arc<Mutex<Valid<Schema>>>,
    index: SchemaIndex,
    allow_mutations: bool,
    leaf_depth: usize,
    minify: bool,
    pub tool: Tool,
}

/// Input for the search tool.
#[derive(JsonSchema, Deserialize)]
pub struct Input {
    /// The search terms
    terms: Vec<String>,
}

/// An error while indexing the GraphQL schema.
#[derive(Debug, thiserror::Error)]
pub enum IndexingError {
    #[error("Unable to index schema: {0}")]
    IndexingError(#[from] apollo_schema_index::error::IndexingError),

    #[error("Unable to lock schema: {0}")]
    TryLockError(#[from] tokio::sync::TryLockError),
}

impl Search {
    pub fn new(
        schema: Arc<Mutex<Valid<Schema>>>,
        allow_mutations: bool,
        leaf_depth: usize,
        index_memory_bytes: usize,
        minify: bool,
    ) -> Result<Self, IndexingError> {
        let root_types = if allow_mutations {
            OperationType::Query | OperationType::Mutation
        } else {
            OperationType::Query.into()
        };
        let locked = &schema.try_lock()?;
        Ok(Self {
            schema: schema.clone(),
            index: SchemaIndex::new(locked, root_types, index_memory_bytes)?,
            allow_mutations,
            leaf_depth,
            minify,
            tool: Tool::new(
                SEARCH_TOOL_NAME,
                format!(
                    "Search a GraphQL schema{}",
                    if minify {
                        " - T=type,I=input,E=enum,U=union,F=interface;s=String,i=Int,f=Float,b=Boolean,d=ID;!=required,[]=list,<>=implements"
                    } else {
                        ""
                    }
                ),
                schema_from_type!(Input),
            ),
        })
    }

    pub async fn execute(&self, input: Input) -> Result<CallToolResult, McpError> {
        let mut root_paths = self
            .index
            .search(input.terms.clone(), Options::default())
            .map_err(|e| {
                McpError::new(
                    ErrorCode::INTERNAL_ERROR,
                    format!("Failed to search index: {e}"),
                    None,
                )
            })?;

        root_paths.truncate(MAX_SEARCH_RESULTS);
        debug!(
            "Root paths for search terms: {}\n{}",
            input.terms.join(", "),
            root_paths
                .iter()
                .map(ToString::to_string)
                .collect::<Vec<String>>()
                .join("\n"),
        );

        let schema = self.schema.lock().await;
        let mut tree_shaker = SchemaTreeShaker::new(&schema);
        for root_path in root_paths {
            let path_len = root_path.inner.len();
            for (i, path_node) in root_path.inner.into_iter().enumerate() {
                if let Some(extended_type) = schema.types.get(path_node.node_type.as_str()) {
                    let (selection_set, depth) = if i == path_len - 1 {
                        (None, DepthLimit::Limited(self.leaf_depth))
                    } else {
                        (
                            path_node.field_name.as_ref().map(|field_name| {
                                vec![Selection::Field(Node::from(Field {
                                    alias: Default::default(),
                                    name: Name::new_unchecked(field_name),
                                    arguments: Default::default(),
                                    selection_set: Default::default(),
                                    directives: Default::default(),
                                }))]
                            }),
                            DepthLimit::Limited(1),
                        )
                    };
                    tree_shaker.retain_type(extended_type, selection_set.as_ref(), depth)
                }
                for field_arg in path_node.field_args {
                    if let Some(extended_type) = schema.types.get(field_arg.as_str()) {
                        // Retain input types with unlimited depth because all input must be given
                        tree_shaker.retain_type(extended_type, None, DepthLimit::Unlimited);
                    }
                }
            }
        }

        let shaken = tree_shaker.shaken().unwrap_or_else(|schema| schema.partial);

        Ok(CallToolResult {
            content: shaken
                .types
                .iter()
                .filter(|(_name, extended_type)| {
                    !extended_type.is_built_in()
                        && schema
                            .root_operation(AstOperationType::Mutation)
                            .is_none_or(|root_name| {
                                extended_type.name() != root_name || self.allow_mutations
                            })
                })
                .map(|(_, extended_type)| {
                    if self.minify {
                        extended_type.minify()
                    } else {
                        extended_type.serialize().to_string()
                    }
                })
                .map(Content::text)
                .collect(),
            is_error: None,
            meta: None,

            // Note: The returned content is treated as text, so no need to structure its output
            structured_content: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rmcp::model::RawContent;
    use rstest::{fixture, rstest};
    use std::ops::Deref;

    const TEST_SCHEMA: &str = include_str!("testdata/schema.graphql");

    fn content_to_snapshot(result: CallToolResult) -> String {
        result
            .content
            .into_iter()
            .filter_map(|c| {
                let c = c.deref();
                match c {
                    RawContent::Text(text) => Some(text.text.clone()),
                    _ => None,
                }
            })
            .collect::<Vec<String>>()
            .join("\n")
    }

    #[fixture]
    fn schema() -> Valid<Schema> {
        Schema::parse(TEST_SCHEMA, "schema.graphql")
            .expect("Failed to parse test schema")
            .validate()
            .expect("Failed to validate test schema")
    }

    #[rstest]
    #[tokio::test]
    async fn test_search_tool(schema: Valid<Schema>) {
        let schema = Arc::new(Mutex::new(schema));
        let search = Search::new(schema.clone(), false, 1, 15_000_000, false)
            .expect("Failed to create search tool");

        let result = search
            .execute(Input {
                terms: vec!["User".to_string()],
            })
            .await
            .expect("Search execution failed");

        assert!(!result.is_error.unwrap_or(false));
        insta::assert_snapshot!(content_to_snapshot(result));
    }

    #[rstest]
    #[tokio::test]
    async fn test_referencing_types_are_collected(schema: Valid<Schema>) {
        let schema = Arc::new(Mutex::new(schema));
        let search = Search::new(schema.clone(), true, 1, 15_000_000, false)
            .expect("Failed to create search tool");

        // Search for a type that should have references
        let result = search
            .execute(Input {
                terms: vec!["User".to_string()],
            })
            .await
            .expect("Search execution failed");

        assert!(!result.is_error.unwrap_or(false));
        assert!(
            content_to_snapshot(result).contains("createUser"),
            "Expected to find the createUser mutation in search results"
        );
    }
}
