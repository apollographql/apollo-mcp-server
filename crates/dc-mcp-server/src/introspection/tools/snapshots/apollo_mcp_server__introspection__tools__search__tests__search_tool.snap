---
source: crates/apollo-mcp-server/src/introspection/tools/search.rs
expression: content_to_snapshot(result)
---
scalar DateTime

enum UserRole {
  ADMIN
  MODERATOR
  USER
  GUEST
}

enum ContentStatus {
  DRAFT
  PUBLISHED
  ARCHIVED
  DELETED
}

type User implements Node {
  role: UserRole!
}

type Post implements Node & Content {
  author: User!
}

type Query {
  user(id: ID!): User
  postsOld(filter: [ID!]): [Post!]! @deprecated(reason: "Use posts instead")
  posts(filter: PostFilter): [Post!]!
}

input PostFilter {
  status: ContentStatus
  authorId: ID
  tags: [String!]
  dateRange: DateRangeInput
}

input DateRangeInput {
  start: DateTime!
  end: DateTime!
}
